engine/services/download_service.cc

```diff
@@ -382,6 +452,44 @@ void DownloadService::RemoveTaskFromStopList(const std::string& task_id) {
   tasks_to_stop_.erase(task_id);
 }
 
+void DownloadService::EmitTaskStarted(const DownloadTask& task) {
+  event_queue_->enqueue(
+      EventType::DownloadEvent,
+      DownloadEvent{.type_ = DownloadEventType::DownloadStarted,
+                    .download_task_ = task});
+}
+
+void DownloadService::EmitTaskStopped(const std::string& task_id) {
+  if (auto it = active_tasks_.find(task_id); it != active_tasks_.end()) {
+    event_queue_->enqueue(
+        EventType::DownloadEvent,
+        DownloadEvent{.type_ = DownloadEventType::DownloadStopped,
+                      .download_task_ = *it->second});
+  }
+}
+
+void DownloadService::EmitTaskError(const std::string& task_id) {
```

Comment:
Why don’t we need lock_guard here but ‘task completed’ we need?

Reply:
I think it's won't be necessary because I'm not needed to update downloadedBytes of the item.